pkg/controllers/elastic_jobset.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"fmt"
+	"strconv"
+
+	batchv1 "k8s.io/api/batch/v1"
+
+	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+)
+
+// jobsToDeleteDownScale gathers the excess jobs during a downscale
+// and deletes the jobs
+func jobsToDeleteDownScale(replicatedJobs []jobset.ReplicatedJob, replicatedJobStatus []jobset.ReplicatedJobStatus, jobItems []batchv1.Job) ([]*batchv1.Job, error) {
+	jobsToDelete := []*batchv1.Job{}
+	for _, replicatedJob := range replicatedJobs {
+		status := findReplicatedJobStatus(replicatedJobStatus, replicatedJob.Name)
+		countOfJobsToDelete := status.Ready - replicatedJob.Replicas
+		if countOfJobsToDelete > 0 {
+			jobsWeDeleted := 0
+			for _, val := range jobItems {
+				if val.Labels[jobset.ReplicatedJobNameKey] != replicatedJob.Name {
+					continue
+				}
+				jobIndex, err := strconv.Atoi(val.Labels[jobset.JobIndexKey])
+				if err != nil {
+					return nil, fmt.Errorf("unable get integer from job index key")
+				}
+				if jobIndex >= int(countOfJobsToDelete) {
```

Comment:
So if we have 4 jobs (indexes 0,1,2,3) and we want to delete 1 job, then `countOfJobsToDelete=1` and we may end up deleting any of job indexes 1,2,3 - so it is non-deterministic and may leave us with an non-contiguous range of job indexes (remaining jobs indexes 0,2,3).

Rather than this, I think we should delete jobs from largest job index to smallest, so the remaining jobs still occupy a contiguous index range - following the same pattern as how the Job controller behaves for elastic indexed jobs, deleting pods from largest to smallest completion indexes).

Reply:
Good point. I went ahead and added this change.

Reply:
@danielvegamyhre , keeping the indices unchanged could be desirable in cases where we don't want to restart the entire jobset right?

Reply:
> @danielvegamyhre , keeping the indices unchanged could be desirable in cases where we don't want to restart the entire jobset right?

Currently we always restart the entire jobset (recreating all child jobs), we don't yet support restarting only a subset of jobs, since there hasn't been a clear use case for that (due to the group-startup semantics required for distributed training, everything must restart, reload from latest checkpoint, do distributed initialization, etc).

What cases are you envisioning where we only want to restart a subset of jobs?